optimade/server/entry_collections/entry_collections.py

```diff
@@ -1,34 +1,87 @@
-from abc import abstractmethod
-from typing import Collection, Tuple, List
+from abc import abstractmethod, ABC
+from typing import Tuple, List, Union
+
+from fastapi import HTTPException
+from lark import Transformer
 
-from optimade.server.mappers import BaseResourceMapper
 from optimade.filterparser import LarkParser
 from optimade.models import EntryResource
-from optimade.server.query_params import EntryListingQueryParams
+from optimade.server.config import CONFIG
+from optimade.server.exceptions import BadRequest
+from optimade.server.mappers import BaseResourceMapper
+from optimade.server.query_params import EntryListingQueryParams, SingleEntryQueryParams
+
 
+class EntryCollection(ABC):
+    """ Backend agnostic base class for collections of EntryResources. """
 
-class EntryCollection(Collection):  # pylint: disable=inherit-non-class
     def __init__(
         self,
         collection,
         resource_cls: EntryResource,
         resource_mapper: BaseResourceMapper,
+        transformer: Transformer,
     ):
         self.collection = collection
         self.parser = LarkParser()
         self.resource_cls = resource_cls
         self.resource_mapper = resource_mapper
+        self.transformer = transformer
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """ Returns the overall number of entries in the collection. """
+
+    @abstractmethod
+    def count(self, **kwargs) -> int:
+        """ Returns the number of entries matching the query specified
+        by the keyword arguments.
+
+        """
+
+    @abstractmethod
+    def find(
+        self, params: EntryListingQueryParams
+    ) -> Tuple[List[EntryResource], int, bool, set]:
+        """
+        Fetches results and indicates if more data is available.
+
+        Also gives the total number of data available in the absence of page_limit.
+
+        Args:
+            params (EntryListingQueryParams): entry listing URL query params
 
-    def __len__(self):
-        return self.collection.count()
+        Returns:
+            Tuple[List[Entry], int, bool, set]: (results, data_returned, more_data_available, fields)
```

Comment:
I have found that since the `mkdocs` plugin for getting API reference links uses square brackets for those links it works better to wrap the "type" here in code-quotations, however that's only if you need to make actual links. For this it might be fine.
But maybe you want to link `data_returned` and `more_data_available` to their model definitions?
```suggestion
            Tuple[List[EntryResource], int, bool, set]: (`results`, `data_returned`, `more_data_available`, `fields`).
```

Reply:
I think there's definitely a trade-off between readable docstrings and super-duper online docs, not sure how far we want to go towards the docs...

Reply:
So again, if I serve this locally, the type is repeated because it is already present in the type hint...

Reply:
Yeah. The mkdocs for `Returns` is terrible. And you can definitely go with your own on this. But `Entry` should definitely be updated to `EntryResource` :)